packages/matter.js/src/protocol/interaction/InteractionServer.ts

```diff
@@ -371,17 +388,248 @@ export function attributePathToId({ endpointId, clusterId, attributeId }: TypeFr
     return `${endpointId}/${clusterId}/${attributeId}`;
 }
 
-function toHex(value: number | undefined) {
-    return value === undefined ? "*" : `0x${value.toString(16)}`;
+export function eventPathToId({ endpointId, clusterId, eventId }: TypeFromSchema<typeof TlvEventPath>) {
+    return `${endpointId}/${clusterId}/${eventId}`;
 }
 
-export class InteractionServer implements ProtocolHandler<MatterDevice> {
+export class EndpointStructure {
+    endpoints = new Map<number, Endpoint>();
+    attributes = new Map<string, (AttributeServer<any> | FabricScopedAttributeServer<any> | FixedAttributeServer<any>)>();
+    attributePaths = new Array<AttributePath>();
+    events = new Map<string, EventServer<any>>();
+    eventPaths = new Array<EventPath>();
+    commands = new Map<string, CommandServer<any, any>>();
+    commandPaths = new Array<CommandPath>();
+
+    public clear() {
+        this.endpoints.clear();
+        this.attributes.clear();
+        this.attributePaths.length = 0;
+        this.events.clear();
+        this.eventPaths.length = 0;
+        this.commands.clear();
+        this.commandPaths.length = 0;
+    }
+
+    public initializeFromEndpoint(endpoint: Endpoint) {
+        this.clear();
+
+        this.verifyAndInitializeStructureElementsFromEndpoint(endpoint); // Initialize Data from Root Endpoint
+        this.initializeStructureFromEndpoints(endpoint); // Initialize Data from Child Endpoints
+    }
+
+    private initializeStructureFromEndpoints(endpoint: Endpoint) {
+        const endpoints = endpoint.getChildEndpoints();
+        for (let endpointIndex = 0; endpointIndex < endpoints.length; endpointIndex++) {
+            this.verifyAndInitializeStructureElementsFromEndpoint(endpoints[endpointIndex]);
+            this.initializeStructureFromEndpoints(endpoints[endpointIndex]);
+        }
+    }
+
+    private verifyAndInitializeStructureElementsFromEndpoint(endpoint: Endpoint) {
+        if (endpoint.id === undefined) {
+            throw new Error(`Endpoint ID is undefined. It needs to be initialized first!`);
+        }
+
+        endpoint.verifyRequiredClusters();
+
+        for (const cluster of endpoint.getAllClusterServers()) {
+            const { id: clusterId, attributes: clusterAttributes, _events: clusterEvents, _commands: clusterCommands } = cluster;
+            // Add attributes
+            for (const name in clusterAttributes) {
```

Comment:
Could do this and following two loops in a generic function.

Reply:
you mean because attrbtes/events/commands are basically "samie-ish". Yes but in fact the "toId" functions and both structures are all different, so wouod all be configurable ... Fund it easier to have "smeish" code :-)

Reply:
But yes good Idea will think tomorrow about it